app/frontend/wallet/shelley/shelley-transaction-planner.ts

```diff
@@ -452,41 +516,24 @@ export const selectMinimalTxPlan = (
     tokenBundle: [],
   }
 
-  // TODO: refactor this
-  for (const utxo of utxos) {
-    inputs.push(utxo)
-    try {
-      const plan = computeTxPlan(inputs, outputs, change, certificates, withdrawals)
-      validateTxPlan(plan)
-      return {
-        success: true,
-        txPlan: plan,
-      }
-    } catch (e) {
-      if (inputs.length === utxos.length) {
-        return {
-          success: false,
-          estimatedFee: computeRequiredTxFee(inputs, outputs, certificates, withdrawals),
-          error: {code: e.name},
-          minimalLovelaceAmount: outputs.reduce(
-            (acc, output) =>
-              output.tokenBundle.length > 0
-                ? acc + computeMinUTxOLovelaceAmount(output.tokenBundle)
-                : 0,
-            0
-          ) as Lovelace,
-        }
-      }
+  const _selectMinimalTxPlan = (
+    numOfInputs: number,
+    allowAddingChangeToFee: boolean
+  ): TxPlanResult => {
+    const inputs: TxInput[] = utxos.slice(0, numOfInputs)
+    const txPlanResult = validateTxPlan(
+      computeTxPlan(inputs, outputs, change, certificates, withdrawals, allowAddingChangeToFee)
+    )
+    if (txPlanResult.success === true) {
+      return txPlanResult
     }
+    return numOfInputs < utxos.length
```

Comment:
do we really need recursion? wouldn't it work as well with an ordinary while loop which is easier to follow?

Regarding the `allowAddingChangeToFee` I'd rather enhance the output of `computeTxPlan` to be able to tell how the fee was obtained, whether any change output has been "burned" - for that I'd consider just adding a `baseFee` property telling the fee attributed to the transaction as such (i.e. based on its base size) and that could be used to infer whether an output has been burned and hence whether it's worth trying to add additional inputs or not

Reply:
I agree, I'm not completely bought into using recursion as well, I can't say I like it very much :smile: but this is rather a simple one (basically just passing an incrementing counter) so I considered it fine. 

I m gonna try the approach you proposed and if It works I am all for it. :smile: